src/main.rs

```diff
@@ -0,0 +1,300 @@
+use axum::ServiceExt;
+
+mod ecs;
+mod errors;
+
+use {
+    axum::{extract::Path, extract::State, http::StatusCode, response::Json, routing::get, Router},
+    bulwark_ext_processor::BulwarkProcessor,
+    clap::{Parser, Subcommand},
+    color_eyre::eyre::Result,
+    envoy_control_plane::envoy::service::ext_proc::v3::external_processor_server::ExternalProcessorServer,
+    errors::*,
+    serde::Serialize,
+    std::net::{IpAddr, Ipv4Addr, SocketAddr},
+    std::{
+        path::PathBuf,
+        sync::{Arc, Mutex},
+    },
+    tokio::task::JoinSet,
+    tonic::transport::Server,
+    tower_http::normalize_path::NormalizePathLayer,
+    tower_layer::Layer,
+    tracing::error,
+    tracing_forest::ForestLayer,
+    tracing_log::LogTracer,
+    tracing_subscriber::layer::SubscriberExt,
+    tracing_subscriber::{EnvFilter, Registry},
+};
+
+/// bulwark-cli launches and interacts with the Bulwark service.
+#[derive(Parser)]
+#[command(author, version, about, long_about = None)]
+struct Cli {
+    /// Log levels: error, warn, info, debug, trace
+    ///
+    /// Default is "info".
+    #[arg(short, long)]
+    log_level: Option<String>,
+
+    /// Log formats: ecs, forest
+    ///
+    /// Default is "ecs".
+    #[arg(short, long)]
+    log_format: Option<String>,
+
+    #[command(subcommand)]
+    command: Option<Commands>,
+}
+
+/// The subcommands supported by the Bulwark CLI.
+#[derive(Subcommand)]
+enum Commands {
+    /// Launch as an Envoy external processor
+    ExtProcessor {
+        /// Sets a custom config file
+        #[arg(short, long, value_name = "FILE")]
+        config: PathBuf,
+    },
+    // TODO: Implement ReverseProxy subcommand
+    // TODO: Implement Check subcommand
+    // TODO: Implement Test subcommand
+    // TODO: Implement Compile subcommand
+}
+
+/// The health state structure tracks the health of the primary service, primarily for the benefit of
+/// external monitoring by load balancers or container orchestration systems. It does not track a
+/// liveness value because this will always be true if the process is running.
+struct HealthState {
+    /// Indicates that the primary service has successfully initialized and is ready to receive requests.
+    /// Once true, it will remain true for the lifetime of the process.
+    started: bool,
+    /// Indicates that the primary service is ready to receive requests.
+    /// In theory, this could become false after the service has started if the system detects that it's
+    /// entered a deadlock state, however this is not currently implemented and does not meaningfully
+    /// differ from the started state.
+    ready: bool,
```

Comment:
I'm mainly using TODOs to track things that I might lose track of. I don't think this is something I'd lose track of, the location of the health check stuff is obvious, but I should probably file a tracking issue for it.

Redis going down might be one realistic source of potential failure, but that's best handled by being resilient to the failure rather than a change to readiness or liveness state. Liveness entering fail state would just kill the Bulwark service and auto-restart it, which won't help anything if Redis is down. Readiness entering fail state might help if Envoy fails open but fail-all-the-way-open seems undesirable and fail closed most likely would make the problem much worse. Best case for a Redis failure would be a future scenario where `Result` return values for the remote state function immediately tell plugins that remote state is unavailable and defer to them to either implement fallback logic, return an unknown decision value, or panic. The main service would remain responsive to Envoy gRPC requests and plugins with no Redis dependency or alternative fallback logic would continue to provide security value.

Deadlock detection is something I want to explore but I'm uncertain how valuable it will be. There's all the infinite possible plugin implementations combined with concurrent plugin execution that make deadlocks seem like something to worry about. But also plugin execution is single-threaded within each guest and I'm not sure killing the whole service is the right answer if a deadlock occurs that's local to a single request context. That seems more like a job for timeouts, which Bulwark already has.